test/application/versioning/Package.test.js

```diff
@@ -0,0 +1,158 @@
+
+      it('reverts', async function () {
+        await assertRevert(this.package.addVersion(version, this.directory_V0.address, { from }))
+      })
+    })
+  })
+
+  describe('getVersion', function () {
+    const version = '0'
+
+    describe('when the requested version was set', function () {
+      beforeEach(async function () {
+        await this.package.addVersion(version, this.directory_V0.address, { from: owner })
+      })
+
+      it('returns the registered directory', async function () {
+        const registeredDirectory = await this.package.getVersion(version)
+        assert.equal(registeredDirectory, this.directory_V0.address)
+      })
+    })
+
+    describe('when the requested version was not set', function () {
+      it('returns the address zero', async function () {
+        const registeredDirectory = await this.package.getVersion(version)
+        assert.equal(registeredDirectory, 0x0)
+      })
+    })
+  })
+
+  describe('hasVersion', function () {
+    const version = '0'
+
+    describe('when the requested version was set', function () {
+      beforeEach(async function () {
+        await this.package.addVersion(version, this.directory_V0.address, { from: owner })
+      })
+
+      it('returns true', async function () {
+        const hasVersion = await this.package.hasVersion(version)
+        assert.isTrue(hasVersion)
+      })
+    })
+
+    describe('when the requested version was not set', function () {
+      it('returns the address zero', async function () {
+        const hasVersion = await this.package.hasVersion(version)
+        assert.isFalse(hasVersion)
+      })
+    })
+  })
+
+  describe('getImplementation', function () {
+    const version = '0'
+    const contract = 'ERC721'
+
+    describe('when the requested version was set', function () {
+      beforeEach(async function () {
+        await this.package.addVersion(version, this.directory_V0.address, { from: owner })
+      })
+
+      describe('when the requested version holds the requested contract name', function () {
+        beforeEach(async function () {
+          await this.directory_V0.setImplementation(contract, implementation_v0, { from: owner })
+        })
+
+        it('returns the requested implementation', async function () {
+          const implementation = await this.package.getImplementation(version, contract)
+          assert.equal(implementation, implementation_v0)
+        })
+      })
+
+      describe('when the requested version does not hold the requested contract name', function () {
+        it('returns the zero address', async function () {
+          const implementation = await this.package.getImplementation(version, contract)
+          assert.equal(implementation, 0x0)
+        })
+      })
+    })
+
+    describe('when the requested version was not set', function () {
+      it('reverts', async function () {
+        await assertRevert(this.package.getImplementation(version, contract))
```

Comment:
To be discussed: I find somehow inconsistent that `ContractDirectory` returns `0x0` when the contract is not defined, but `Package` reverts if the version isn't.

Reply:
Registered here https://github.com/zeppelinos/upgradeability-lib/issues/48